### PR TITLE
SpacingSizesControl: Fix problem with the slider position being reset when saving global styles

### LINKED ISSUE #50940: Styles sidebar: Site logo spacing controls are reset on saving
### Description

When I try to use a global styles preset for margin for the site logo block, the control in the styles sidebar is reset when I save.

Update: same with padding.

### Step-by-step reproduction instructions

1. Activate a block theme. I used Twenty Twenty-Three.
2. Add a site logo block with an image.
3.  In the Site Editor, Styles sidebar, select the site logo block.
4. In the dimensions panel, select a spacing preset. Either one for all sides, or individual size.
5. Save the style changes
6.  See that the control is reset, but in the preview in the site logo panel, the margin is applied.

### Screenshots, screen recording, code snippet

https://github.com/WordPress/gutenberg/assets/7422055/f546ee7a-4a26-43d6-812e-c99064ddd7cc

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by t-hamano**
Fixes: #50940

## What?
This PR fixes a problem with the `SpacingSizesControl` slider position being reset when the global style is saved.

https://github.com/WordPress/gutenberg/assets/54422211/1875ec50-7b57-4ebd-8d04-e5c17678087a

## Why?

I will try to explain the cause of this in turn.

When you move a slider in the `SpacingSizesControl` component, it holds an object with a CSS variable containing a `pipe` and a `colon` as values, like this:

```
{
  "top": "var:preset|spacing|50",
  "right": "var:preset|spacing|50",
  "bottom": "var:preset|spacing|50",
  "left": "var:preset|spacing|50"
}
```

If you save the global style after moving the slider, an object with the usual CSS variables will be passed to the `DimensionsPanel` component as a value, like this:

```
{
  "top": "var(--wp--preset--spacing--50)",
  "right": "var(--wp--preset--spacing--50)",
  "bottom": "var(--wp--preset--spacing--50)",
  "left": "var(--wp--preset--spacing--50)"
}
```

Upon receiving the value, the `DimensionsPanel` component attempts to decode the normal CSS variable into a CSS variable containing the `pipe` and `colon` by means of the `decodeValue()` function.

For Padding, here it is.

https://github.com/WordPress/gutenberg/blob/f10f85f5779eab8a9daeb00ef6721543dce91201/packages/block-editor/src/components/global-styles/dimensions-panel.js#L258

However, the `getValueFromVariable()` function that the `decodeValue()` function calls internally expects the passed value to be of type `string` and returns the received value as is if the condition is not met:

https://github.com/WordPress/gutenberg/blob/f10f85f5779eab8a9daeb00ef6721543dce91201/packages/block-editor/src/components/global-styles/utils.js#L307

Since this value is not a CSS variable containing `pipe` and `colon`, the `SpacingSizesControl` component that receives the value will assume it is not a preset value:

https://github.com/WordPress/gutenberg/blob/f10f85f5779eab8a9daeb00ef6721543dce91201/packages/block-editor/src/components/spacing-sizes-control/utils.js#L17

As a result, the slider position will be lost.

## How?

We need to decode the value of each of the objects with `top` / `right` / `bottom` / `left` properties in some way.

I found #48070 where this component was affected, but I expect that perhaps before this PR, the object type was also taken into account when decoding.

I have added a condition and processing to check if the value is an object in the `decodeValue()` function, but there might be a better way.

## Testing Instructions

- Open the global styles to access blocks with Dimensions controls, such as the paragraph block.
- Manipulate the slider.
- Make sure the slider position is not lost after saving.

### CODE REVIEW

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for the fix @t-hamano 👍 

This is testing well. I could replicate the issue on trunk and this PR fixes it for me.

> I found https://github.com/WordPress/gutenberg/pull/48070 where this component was affected, but I expect that perhaps before this PR, the object type was also taken into account when decoding.

A git bisect pointed to a different PR that might be the cause: https://github.com/WordPress/gutenberg/pull/50366

That PR changed the format of the preset values returned from global styles, so it makes sense that it caused the issue with the spacing control.

It's possible that any other controls testing for the preset values might have been affected. I didn't spot any odd behaviour though testing things like font sizes, font families and the like.

I'm happy to approve this one but feel free to get a second review.

**Comment by t-hamano** [comment]
@aaronrobertshaw 
Thank you for the detailed investigation and testing of the cause! I would like to merge this PR, but would like to submit a separate issue if I find problems with other controls.

**Comment by carolinan** [comment]
thank you!

**Comment by andrewserong** [comment]
I think this PR introduced an error with some blocks at the block level in global styles, when the `rawValue` is `null` which unfortunately evaluates to the type of `object`. Sorry for not catching this sooner! I noticed that accessing the Columns screen crashes the editor for me, and have a fix up over in #51252.

**Pull request merged**
